Adding an Example: Create Map_with_cartopy_projections.py

mixstam1821 opened the pull request:
Patches are transformed according to the cartopy projections resulting in a map with custom projection in Bokeh. Useful for representations of the global distribution of various climatic parameters.

<img width="796" alt="Screenshot 2025-01-27 at 19 33 46" src="https://github.com/user-attachments/assets/709c98fa-bae7-4727-af49-8d1f0ed59e80" />

mixstam1821 committed:
Create Map_with_cartopy_projections.py

Patches are transformed according to the cartopy projections resulting in a map with custom projection in Bokeh. Useful for representations of the global distribution of various climatic parameters.

bryevdv commented:
@mixstam1821 Thanks for sharing this example and for submitting the PR! This is a really neat example i would very much like to eventually get something very much like this added. I think in it's current form it is a jumping off point. A few quick comments:

* Some repo changes will be needed to add `cartopy` and `shapely` to the test envs so that the example can be run in CI
* This example should live under [`examples/topics/geo`](https://github.com/bokeh/bokeh/tree/branch-3.7/examples/topics/geo) rather than `examples/plotting`
* The example should also get added to the gallery, and get a section in the User's Guide [Geographical data chapter](https://docs.bokeh.org/en/latest/docs/user_guide/topics/geo.html)

We can help with those things. But first, a question mostly for @bokeh/dev: What if any of the helper code should we consider adding to a `bokeh.geo` module or similar? This example, as it stands, is a bit longer and more complicated than we typically go for in example scripts. Generating the patches or the coastlines is not "hard" as seen above, but having to do it by hand would probably put off many users. Any helpers functions would be up-front APIs (i.e. data pre-processing in Python, no "updates" on the JS side) similar to the `hexbin` helpers. But I think if this kind of plot was dead simple to achieve in Bokeh, it would actually get a lot of traction. We can of course, merge a more complicated version of the script and punt the question of making it easer out of the box down the road. 

Edit: Another example of something we could put in a `bokeh.geo` module would be some simple helpers for web mercator that are probably a decade overdue...

bryevdv commented:
Noting that one line changed generates the Mollweide version (ignore the title) that is actually really cool...

<img width="825" alt="Screenshot 2025-01-27 at 19 34 34" src="https://github.com/user-attachments/assets/f76a4377-fdc1-4d7e-9e23-380ce42a9af5" />

mattpap commented on examples/plotting/Map_with_cartopy_projections.py:
```diff
@@ -0,0 +1,128 @@
+# Robinson
+import numpy as np
+import cartopy.crs as ccrs
+import cartopy.feature as cfeature
```
`cartopy` has to be added to `conda/environment-test-*.yml` under `# tests` section.

bryevdv replied:
`shapely` as well, I think.

mattpap commented on examples/plotting/Map_with_cartopy_projections.py:
```diff
@@ -0,0 +1,128 @@
+                    segment = normalized_coords[valid_indices]
+                    
+                    # Transform coordinates
+                    tt = projection.transform_points(ccrs.PlateCarree(), 
+                                                    segment[:, 0], 
+                                                    segment[:, 1])
```
Let me suggest this:
```suggestion
                    tt = projection.transform_points(
                        ccrs.PlateCarree(), 
                        segment[:, 0], 
                        segment[:, 1],
                    )
```
formatting here and in all other cases. Easier to add new arguments and when refactoring, you don't have to adjust spacing in multiple lines.

mattpap replied:
Also, there are other relevant stylistic changes collected e.g. [here](https://github.com/bokeh/bokeh/actions/runs/13000275400/job/36257445220?pr=14258). You can verify locally by running `pytest tests/codebase`, assuming you have your environment [set up](https://docs.bokeh.org/en/latest/docs/dev_guide/setup.html) for development.

mattpap commented:
> What if any of the helper code should we consider adding to a `bokeh.geo` module or similar? This example, as it stands, is a bit longer and more complicated than we typically go for in example scripts. 

I'm not sure what other, but this alone would be quite valuable to have.

mixstam1821 committed:
Merge branch 'bokeh:branch-3.7' into branch-3.7

mixstam1821 committed:
Create Nereus_countries_explorer.py

A demo with the Bokeh server to explore the countries of the world along with some useful pieces of information.

bryevdv commented:
Hi @mixstam1821 let's keep each PR focused to a single example for now. Can you remove the second app example from this branch? And as a concrete first step, move the cartopy example to `examples/topics/geo/projected_colormap.py` to be more in line with our typical naming style?

bryevdv commented:
@mixstam1821 can you update to let us know whether you can continue to iterate on this PR? If not that's fine but then we will close and then re-open a new PR with the code moved and updated, and with additional docs updates, etc.

mixstam1821 commented:
> @mixstam1821 can you update to let us know whether you can continue to iterate on this PR? If not that's fine but then we will close and then re-open a new PR with the code moved and updated, and with additional docs updates, etc.

Hi @bryevdv, I'm currently busy and not very familiar with the GitHub process yet. Feel free to proceed with closing and reopening a new PR with the necessary updates. Thanks for the understanding. I hope in the future to be more ready for future PRs.

mixstam1821 committed:
Delete examples/server/app/Nereus_countries_explorer.py

bryevdv commented:
Thanks @mixstam1821 !

mixstam1821 committed:
Merge branch 'bokeh:branch-3.7' into branch-3.7

mosc9575 commented:
I am closing this for #14548. I try to reuse the most of the code.
